shadowsocks/udp.go

```diff
@@ -187,45 +188,60 @@ func (s *udpService) Stop() error {
 }
 
 type natentry struct {
-	conn net.PacketConn
+	Conn net.PacketConn
 	// We store the client location in the NAT map to avoid recomputing it
 	// for every outbound packet in a UDP-based connection.
 	clientLocation string
+	// Number of upstream packets sent on Conn.
+	count int64
+	// Number of DNS query packets sent on Conn.
+	countDNS int64
 }
 
 // Packet NAT table
 type natmap struct {
 	sync.RWMutex
-	keyConn map[string]natentry
+	keyConn map[string]*natentry
 	timeout time.Duration
 	metrics metrics.ShadowsocksMetrics
 }
 
 func newNATmap(timeout time.Duration, sm metrics.ShadowsocksMetrics) *natmap {
 	m := &natmap{metrics: sm}
-	m.keyConn = make(map[string]natentry)
+	m.keyConn = make(map[string]*natentry)
 	m.timeout = timeout
 	return m
 }
 
-func (m *natmap) Get(key string) (net.PacketConn, string) {
+func (m *natmap) Get(key string) *natentry {
 	m.RLock()
 	defer m.RUnlock()
-	entry := m.keyConn[key]
-	return entry.conn, entry.clientLocation
+	return m.keyConn[key]
 }
 
 // Refresh the NAT mapping.  This should be called on every write for
 // outbound-refresh behavior (as required by RFC 4787 Section 4.3).
-func (m *natmap) Refresh(targetConn net.PacketConn) {
-	targetConn.SetReadDeadline(time.Now().Add(m.timeout))
+func (m *natmap) Refresh(entry *natentry, dstPort int) {
+	count := atomic.AddInt64(&entry.count, 1)
+	timeout := m.timeout
+	if dstPort == 53 {
+		// This is a DNS query.
+		countDNS := atomic.AddInt64(&entry.countDNS, 1)
+		if countDNS == count {
```

Comment:
Won't this always be equal since the dstPort is constant for an association?

Reply:
Also, do we need those counters?

Reply:
The source port is constant for an association, but `dstPort` (maybe I should call it `targetPort`?) is not guaranteed to be constant.  This code is very conservative: clients can send any mixture of packets on any association, and the NAT behavior should be correct.

In truth, I'm not aware of any situation in which a client would use a single association for both DNS and non-DNS traffic.  If we're willing to assume that an association that starts with a packet to port 53 remains DNS-only, this code could be a lot simpler (and also more efficient).  For example, we wouldn't need the counters, as you noted.

Reply:
You are right, each UDP association is client IP:port <-> proxy port. I confused things. And yeah, one could potentially use the association to send packets to different targets (even different IPs).

I think calling it `targetPort` will be helpful.

Here is a proposal:

If you take the wrapper approach:
On WriteTo, set the timeout to `max(current timeout, now() + 17s)` if the destination is port 53, or else set it to `max(current timeout, now() + 5m)` (This could be without the max, but using the max is more robust - a refresh should never shorten the timeout).

Reply:
It seems we don't have access to the Deadline that was set, but we could save that in the wrapper.

Reply:
OK, I implemented a wrapper.  Rather than duplicate the deadline, I added a mode flag that allows us to set the right deadline and implement fast close.